Cylindrical transformations with well-defined phi angle

jngrad opened the pull request:
Follow-up to #4014, closes #4019

Description of changes:
- rewrite coordinate transformations with matrix operations

jngrad committed:
utils: Calculate the angle between vectors

jngrad committed:
tests: Add floating point tolerance

jngrad committed:
utils: Refactor coordinate transformation code

jngrad commented:
@KaiSzuttor could you please have a look at the failing tests with cylindrical LB and `HollowConicalFrustum` in commit 764cc32e8f8? Several tests failed because `[1, 0, 0]` was used as both the phi0 axis and cylinder longitudinal axis, which lead to an under-defined phi angle, and thus to a division by zero in the matrix inversion, hence NaN values. But after removing the lines in question, a few tests still fail with NaN values.

KaiSzuttor commented on src/core/observables/CylindricalLBProfileObservable.hpp:
```diff
@@ -47,16 +47,15 @@ class CylindricalLBProfileObservable : public CylindricalProfileObservable {
       // z-axis symmetry.
-      std::tie(theta, rotation_axis) =
-          Utils::rotation_params(Utils::Vector3d{{0.0, 0.0, 1.0}}, axis);
+      constexpr Utils::Vector3d z_axis{{0.0, 0.0, 1.0}};
```
why do we still manually rotate at the call site? I thought the idea was to include that into the transformation? Then you can also test it properly.

jngrad committed:
tests: Comment out testcases with under-defined phi

The cylindrical transformations takes by default the x-axis as the
orientation vector, leading to under-defined phi values when the
x-axis is selected as the longitudinal axis.

jngrad commented:
Alright, so the issues in the testsuite come from test cases where the x-axis is used both for the orientation and the longitudinal axis, which lead to an under-defined phi angle (the rotation matrix is not invertible when two columns are identical). They are temporarily commented out in 90d21a85510b7.

jngrad commented:
Multiple failures in CI for 90d21a85, each with a different segfault backtrace. Cannot reproduce locally in Release and Debug builds.

christophlohrmann commented:
I am trying to get this PR to work, but I have a quick question: Should we assert that the ``axis`` and the phi=0 ``orientation`` are orthogonal or should we just take the orthogonal component in case there is a tilt?

KaiSzuttor commented:
i'm in favour of asserting

christophlohrmann commented:
In that case we should make it such that you have to provide ``orientation`` if you use ``axis``

KaiSzuttor commented:
can this be closed because of #4114?
